serenity-model/src/main/java/net/thucydides/core/requirements/FileSystemRequirementsTagProvider.java

```diff
@@ -130,29 +130,31 @@ public FileSystemRequirementsTagProvider(String rootDirectory) {
      */
     public List<Requirement> getRequirements() {
         if (requirements == null) {
-            requirementsLock.lock();
-            try {
-                Set<Requirement> allRequirements = new HashSet();
-                Set<String> directoryPaths = getRootDirectoryPaths();
-
-                for (String path : directoryPaths) {
-                    File rootDirectory = new File(path);
-                    logger.trace("Loading requirements from {}", rootDirectory);
-                    if (rootDirectory.exists()) {
-                        allRequirements.addAll(loadCapabilitiesFrom(rootDirectory.listFiles(thatAreFeatureDirectories())));
-                        allRequirements.addAll(loadStoriesFrom(rootDirectory.listFiles(thatAreStories())));
+            synchronized (requirementsLock) {
+                if (requirements == null) {
+                    try {
+                        Set<Requirement> allRequirements = new HashSet<>();
+                        Set<String> directoryPaths = getRootDirectoryPaths();
+
+                        for (String path : directoryPaths) {
+                            File rootDirectory = new File(path);
+                            logger.trace("Loading requirements from {}", rootDirectory);
+                            if (rootDirectory.exists()) {
+                                allRequirements.addAll(loadCapabilitiesFrom(rootDirectory.listFiles(thatAreFeatureDirectories())));
+                                allRequirements.addAll(loadStoriesFrom(rootDirectory.listFiles(thatAreStories())));
+                            }
+                        }
+                        requirements = new ArrayList<>(allRequirements);
+                        Collections.sort(requirements);
+                    } catch (IOException e) {
+                        requirements = NO_REQUIREMENTS;
+                        throw new IllegalArgumentException("Could not load requirements from '" + rootDirectory + "'", e);
                     }
+                    requirements = addParentsTo(requirements);
                 }
-                requirements = new ArrayList<>(allRequirements);
-                Collections.sort(requirements);
-            } catch (IOException e) {
-                requirements = NO_REQUIREMENTS;
-                throw new IllegalArgumentException("Could not load requirements from '" + rootDirectory + "'", e);
             }
-            requirements = addParentsTo(requirements);
-            requirementsLock.unlock();
         }
-        return requirements;
+        return Collections.unmodifiableList(requirements);
```

Comment:
I would maybe just retain this line and see if this improves things.

Reply:
No, it is IMO broken without DCL, because the requirements reference could escape the 'guarded' code region